sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs

```diff
@@ -0,0 +1,258 @@
+            }
+            else
+            {
+                // Check for "this.foo"
+                if (strippedExpression is MemberAccessExpressionSyntax member &&
+                    member.Expression is ThisExpressionSyntax thisExpression &&
+                    semanticModel.GetSymbolInfo(expression).Symbol is IFieldSymbol field2)
+                {
+                    return new FieldData(accessorKind, field2, member.Name.GetLocation());
+                }
+            }
+
+            return null;
+        }
+
+        private struct PropertyData
+        {
+            public PropertyData(IPropertySymbol propertySymbol, FieldData? returned, FieldData? updated)
+            {
+                this.PropertySymbol = propertySymbol;
+                this.FieldReturned = returned;
+                this.FieldUpdated = updated;
+            }
+
+            public IPropertySymbol PropertySymbol { get; }
+
+            public FieldData? FieldReturned { get; }
+
+            public FieldData? FieldUpdated { get; }
+        }
+
+        private enum AccessorKind
+        {
+            Getter,
+            Setter
+        }
+
+        private struct FieldData
+        {
+            public FieldData(AccessorKind accessor, IFieldSymbol field, Location location)
+            {
+                this.AccessorKind = accessor;
+                this.Field = field;
+                this.Location = location;
+            }
+
+            public AccessorKind AccessorKind { get; }
+
+            public IFieldSymbol Field { get; }
+
+            public Location Location { get; }
+        }
+
+        /// <summary>
+        /// The rule decides if a property is returning/settings the expected field.
+        /// We decide what the expected field name should be based on a fuzzy match
+        /// between the field name and the property name.
+        /// This class hides the details of matching logic.
+        /// </summary>
+        private class PropertyToFieldMatcher
+        {
+            private readonly IDictionary<IFieldSymbol, string> fieldToStandardNameMap;
+
+            public PropertyToFieldMatcher(IEnumerable<IFieldSymbol> fields)
+            {
+                // Calcuate and cache the standardised versions of the field names to avoid
+                // calculating them every time
+                this.fieldToStandardNameMap = fields.ToDictionary(f => f, f => GetCanonicalFieldName(f.Name));
+            }
+
+            public IFieldSymbol GetSingleMatchingFieldOrNull(IPropertySymbol propertySymbol)
+            {
+                // We're not caching the property name as only expect to be called once per property
+                var standardisedPropertyName = GetCanonicalFieldName(propertySymbol.Name);
+
+                var matchingFields = fieldToStandardNameMap.Keys
+                    .Where(k => AreCanonicalNamesEqual(fieldToStandardNameMap[k], standardisedPropertyName));
+
+                if (matchingFields.Count() != 1)
```

Comment:
[cosmetic] you could do `ToList()` on line 242 and then check `Count`. This way executing `Single()` at the end will not check the first item one more time.